### PR TITLE
feat: adds date_format function

### LINKED ISSUE #3139: feat: date_format(datetime, format) to format a timestamp/date/datetime value
### What problem does the new feature solve?

We need a function `date_format` to format a timestamp, date or datetime value into a string value.

For example:

```sql
select date_format(now(), ,'%Y-%m-%d %T.%000');
```

### What does the feature do?

Format a format a timestamp/date/datetime value into string.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by killme2008**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Main changes:

* Adds `date_format` function to format timestamp/date/datetime by the given format.
* todo:
  * aware of timezone setting

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [ ]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)

Close #3139

### CODE REVIEW

File: src/common/function/src/scalars/date/date_format.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+use datatypes::prelude::{ConcreteDataType, MutableVector, ScalarVectorBuilder};
+use datatypes::vectors::{StringVectorBuilder, VectorRef};
+use snafu::{ensure, ResultExt};
+
+use crate::function::{Function, FunctionContext};
+use crate::helper;
+
+/// A function that formats timestamp/date/datetime into string by the format
+#[derive(Clone, Debug, Default)]
+pub struct DateformatFunction;
```
Comment (sunng87) [inline src/common/function/src/scalars/date/date_format.rs:29]: DateFormatFunction

File: src/common/function/src/function.rs [hunk 2]
```diff
@@ -15,21 +15,22 @@
...
 use common_query::prelude::Signature;
+use common_time::timezone::get_timezone;
+use common_time::Timezone;
 use datatypes::data_type::ConcreteDataType;
 use datatypes::vectors::VectorRef;
 
 #[derive(Clone)]
 pub struct FunctionContext {
-    pub tz: Tz,
+    pub timezone: Timezone,
```
Comment (sunng87) [inline src/common/function/src/function.rs:27]: Is it possible to access `QueryContext` from UDF?

Comment (killme2008) [reply src/common/function/src/function.rs:27]: @waynexia already has a solution, but no time to implement 
image

Comment (sunng87) [reply src/common/function/src/function.rs:27]: Discussed with datafusion developer, the implementation of built-in `now` can be referenced: https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/functions.rs#L499

Comment (killme2008) [reply src/common/function/src/function.rs:27]: After discussing with @waynexia , we have a temporal solution. Finally, we hope that Datafusion could change the API that can accept a context for functions while evaluating.

**Review by sunng87 — APPROVED** [review]

File: src/common/time/src/timestamp.rs [hunk 3]
```diff
@@ -290,32 +292,50 @@ impl Timestamp {
...
     pub fn to_local_string(&self) -> String {
+        // Safety: the format is valid
         self.as_formatted_string("%Y-%m-%d %H:%M:%S%.f", None)
+            .unwrap()
     }
 
     /// Format timestamp for given timezone.
     /// If `tz==None`, the server default timezone will used.
-    pub fn to_timezone_aware_string(&self, tz: Option<Timezone>) -> String {
+    pub fn to_timezone_aware_string(&self, tz: Option<&Timezone>) -> String {
```
Comment (waynexia) [inline src/common/time/src/timestamp.rs:309]: should we also include timezone info in those formatted string?

**Review by waynexia — APPROVED** [review]
LGTM

File: src/common/time/src/timestamp.rs (hunk 3 above)
Comment (killme2008) [reply src/common/time/src/timestamp.rs:309]: I'm not sure, @v0y4g3r please take a look.

Comment (killme2008) [reply src/common/time/src/timestamp.rs:309]: I found it's only used in MySQL protocol.

**Pull request merged**